File: backend/app/services/sale_discount.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal

from app.services.discount_calculator import money


SALE_DISCOUNT_TYPES = {"PERCENTAGE", "FIXED_AMOUNT"}


@dataclass(frozen=True)
class SaleDiscountError(ValueError):
    code: str
    message: str


def calculate_sale_discount(subtotal: Decimal, discount_type: str, discount_value: Decimal) -> Decimal:
    """Calculate a checkout-level discount using Decimal currency rounding.

    The caller supplies the subtotal derived from currently locked product prices;
    a client-provided total is deliberately never used here.
    """
    subtotal = money(subtotal)
    if discount_type not in SALE_DISCOUNT_TYPES:
        raise SaleDiscountError("DISCOUNT_TYPE_INVALID", "Select percentage or fixed-amount discount.")
    if discount_type == "PERCENTAGE":
        if discount_value < Decimal("0") or discount_value > Decimal("100"):
            raise SaleDiscountError("DISCOUNT_PERCENTAGE_INVALID", "Discount percentage must be between 0 and 100.")
        return money(subtotal * discount_value / Decimal("100"))
    if discount_value < Decimal("0"):
        raise SaleDiscountError("DISCOUNT_AMOUNT_INVALID", "Discount amount cannot be negative.")
    if discount_value > subtotal:
        raise SaleDiscountError("DISCOUNT_AMOUNT_INVALID", "Discount amount cannot be greater than the subtotal.")
    return money(discount_value)

File: frontend/src/pages/saleDiscount.ts
export type SaleDiscountType = "PERCENTAGE" | "FIXED_AMOUNT";

export type SaleDiscountPreview = {
  valid: boolean;
  error: string;
  discountAmount: number;
  total: number;
};

// Checkout totals are only a live preview. The backend repeats this calculation
// using Decimal values from locked current prices before it commits the sale.
const currency = (value: number) => Math.round((value + Number.EPSILON) * 100) / 100;

export function previewSaleDiscount(subtotal: number, discountType: SaleDiscountType, rawValue: string): SaleDiscountPreview {
  const value = rawValue.trim();
  if (!/^\d+(?:\.\d{0,2})?$/.test(value)) return { valid: false, error: "Enter a valid discount value.", discountAmount: 0, total: currency(subtotal) };
  const discountValue = Number(value);
  if (!Number.isFinite(discountValue)) return { valid: false, error: "Enter a valid discount value.", discountAmount: 0, total: currency(subtotal) };
  if (discountType === "PERCENTAGE") {
    if (discountValue < 0 || discountValue > 100) return { valid: false, error: "Discount percentage must be between 0 and 100.", discountAmount: 0, total: currency(subtotal) };
    const discountAmount = currency(subtotal * discountValue / 100);
    return { valid: true, error: "", discountAmount, total: currency(subtotal - discountAmount) };
  }
  if (discountValue < 0 || discountValue > subtotal) return { valid: false, error: discountValue > subtotal ? "Discount amount cannot be greater than the subtotal." : "Discount amount cannot be negative.", discountAmount: 0, total: currency(subtotal) };
  const discountAmount = currency(discountValue);
  return { valid: true, error: "", discountAmount, total: currency(subtotal - discountAmount) };
}

export function saleDiscountLabel(discountType: SaleDiscountType, value: string) {
  return discountType === "PERCENTAGE" ? `Discount (${value || "0"}%)` : "Discount";
}
